### PR TITLE
[Spark] Test type widening compatibility with other Delta features

### PR DESCRIPTION
**Pull request by johanl-db**
## Description
Additional tests covering type widening and:
- Reading CDF
- Column mapping
- Time travel
- RESTORE
- CLONE

## How was this patch tested?
Test only

### CODE REVIEW

File: spark/src/test/scala/org/apache/spark/sql/delta/DeltaTypeWideningSuite.scala [hunk 1]
```diff
@@ -505,6 +506,84 @@ trait DeltaTypeWideningNestedFieldsTests {
...
+    assert(readDeltaTable(tempPath).schema("b").dataType === ShortType)
+    checkAnswer(sql(s"SELECT b FROM delta.`$tempPath`"), Seq(Row(1), Row(2)))
+    append(Seq(3).toDF("b").select($"b".cast(ShortType)))
+    checkAnswer(sql(s"SELECT b FROM delta.`$tempPath`"), Seq(Row(1), Row(2), Row(3)))
+
+    // Change column type again, add more data.
+    sql(s"ALTER TABLE delta.`$tempPath` CHANGE COLUMN b TYPE int")
+    assert(readDeltaTable(tempPath).schema("b").dataType === IntegerType)
+    append(Seq(4).toDF("b"))
+    checkAnswer(sql(s"SELECT b FROM delta.`$tempPath`"), Seq(Row(1), Row(2), Row(3), Row(4)))
```
Comment (tomvanbussel) [inline spark/src/test/scala/org/apache/spark/sql/delta/DeltaTypeWideningSuite.scala:568]: Do we also have a test for dropping the feature with a column that was renamed after its type was changed?

File: spark/src/test/scala/org/apache/spark/sql/delta/DeltaTypeWideningSuite.scala [hunk 2]
```diff
@@ -1252,6 +1399,50 @@ trait DeltaTypeWideningTableFeatureTests {
...
+  test("dropping feature after RESTORE correctly rewrite files with old type") {
+    addSingleFile(Seq(1), ShortType)
+    addSingleFile(Seq(2), ShortType)
+    sql(s"ALTER TABLE delta.`$tempPath` CHANGE COLUMN a TYPE INT")
+    // Delete the first file which will be added back again with RESTORE.
+    sql(s"DELETE FROM delta.`$tempPath` WHERE a = 1")
+    addSingleFile(Seq(Int.MinValue), IntegerType)
+
+    // RESTORE to version 2 -> ALTER TABLE CHANGE COLUMN TYPE.
+    // The type change is then still present and the first file initially added at version 0 is
```
Comment (tomvanbussel) [inline spark/src/test/scala/org/apache/spark/sql/delta/DeltaTypeWideningSuite.scala:1433]: What will happen when restoring to before the type change? We should still require history truncation in this case.

**Review by tomvanbussel — COMMENTED** [review]
Added a few suggestions for additional tests. Do we already have tests for constraints and generated columns. AFAIK we disallow type changes on generated columns, but constraints might be interesting as it's possible to create a constraint after changing the type.

File: spark/src/test/scala/org/apache/spark/sql/delta/DeltaTypeWideningSuite.scala [hunk 3]
```diff
@@ -505,6 +507,84 @@ trait DeltaTypeWideningNestedFieldsTests {
...
+      parameters = Map(
+        "start" -> "1",
+        "end" -> "1",
+        "readSchema" -> ".*",
+        "readVersion" -> "3",
+        "incompatibleVersion" -> "1",
+        "config" -> ".*defaultSchemaModeForColumnMappingTable"
+      )
+    )
+    checkAnswer(readCDF(start = 3, end = 3), Seq(Row(3, "insert"), Row(4, "insert")))
```
[the remark this answers is not in the document]
Comment (johanl-db) [reply spark/src/test/scala/org/apache/spark/sql/delta/DeltaTypeWideningSuite.scala:645]: You can directly provide a snapshot to use to get the read schema, I added a test for this.

File: spark/src/test/scala/org/apache/spark/sql/delta/DeltaTypeWideningSuite.scala (hunk 2 above)
Comment (johanl-db) [reply spark/src/test/scala/org/apache/spark/sql/delta/DeltaTypeWideningSuite.scala:1483]: I added a test to cover this. Dropping the feature will indeed fail as historical versions still contain the feature, asking to wait for the retention period to expire

**Comment by johanl-db** [comment]
> Added a few suggestions for additional tests. Do we already have tests for constraints and generated columns. AFAIK we disallow type changes on generated columns, but constraints might be interesting as it's possible to create a constraint after changing the type.

There are tests already for generated columns and constraints, added as part of https://github.com/delta-io/delta/pull/2881

I added a test here to cover constraints + RESTORE which is an interesting case:
1. Change column `a` type from `byte` to `int`
2. Add a CHECK constraint on column `a`.
3. RESTORE table to before the type change.

The constraint was added on type `int` and the column now has type `byte`. This does work though because CHECK constraints are part of the table metadata and it gets removed during RESTORE.

**Review by tomvanbussel — APPROVED** [review]
LGTM!

**Pull request merged**